src/backend/libs-and-metadata.md

```diff
@@ -0,0 +1,192 @@
+[codegen unit]: ../backend/codegen.md
+[embed-bitcode]: https://doc.rust-lang.org/rustc/codegen-options/index.html#embed-bitcode
+[linker-plugin-lto]: https://doc.rust-lang.org/rustc/codegen-options/index.html#linker-plugin-lto
+
+### dylib
+
+A `dylib` is a platform-specific shared library. It includes the `rustc`
+[metadata] in a special section called `.rustc` in a compressed format.
+
+### rmeta
+
+An `rmeta` file is custom binary format that contains the [metadata] for the
+crate. This file can be used for fast "checks" of a project by skipping all
+code generation (as is done with `cargo check`), collecting enough information
+for documentation (as is done with `cargo doc`), or for
+[pipelining](#pipelining). This file is created if the
+[`--emit=metadata`][emit] CLI option is used.
+
+`rmeta` files do not support linking, since they do not contain compiled
+object files.
+
+[emit]: https://doc.rust-lang.org/rustc/command-line-arguments.html#option-emit
+
+## Metadata
+
+The metadata contains a wide swath of different elements. This guide will not
+go into detail of every field it contains. You are encouraged to browse the
+[`CrateRoot`] definition to get a sense of the different elements it contains.
+Everything about metadata encoding and decoding is in the [`rustc_metadata`]
+package.
+
+Here are a few highlights of things it contains:
+
+* The version of the `rustc` compiler. The compiler will refuse to load files
+  from any other version.
+* The [Strict Version Hash](#strict-version-hash) (SVH). This helps ensure the
+  correct dependency is loaded.
+* The [Crate Disambiguator](#crate-disambiguator). This is a hash used for
+  various things to disambiguate between different crates of the same name.
+* Information about all the source files in the library. This can be used for
+  a variety of things, such as diagnostics pointing to sources in a
+  dependency.
+* Information about exported macros, traits, types, and items. Generally,
+  anything that's needed to be known when a path references something inside a
+  crate dependency.
+* Encoded [MIR]. This is optional, and only encoded if needed for code
+  generation. `cargo check` skips this for performance reasons.
+
+[`CrateRoot`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/rmeta/struct.CrateRoot.html
+[`rustc_metadata`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/index.html
+[MIR]: ../mir/index.md
+
+### Strict Version Hash
+
+The Strict Version Hash ([SVH], also known as the "crate hash") is a 64-bit
+hash that is used to ensure that the correct crate dependencies are loaded. It
+is possible for a directory to contain multiple copies of the same dependency
+built with different settings, or built from different sources. The crate
+loader will skip any crates that have the wrong SVH.
+
+The SVH is also used for the [incremental compilation] session filename,
+though that usage is mostly historic.
```

Comment:
What's used instead? I'm not sure what you mean by historic here.

Reply:
In the past, `rustc` would use the incremental directories of dependencies for resolving queries, and would use the SVH in the directory name to ensure it was querying the correct dependency. That was removed (it no longer looks for dependencies), and AFAIK it no longer looks at the SVH in the filename.  I think it could be replaced with a random number and have the same effect (it uses a random number before it knows the svh, it gets renamed at the end).  

The incremental naming convention is described at https://doc.rust-lang.org/nightly/nightly-rustc/rustc_incremental/persist/fs/index.html.